harvester/records/validators.py

```diff
@@ -0,0 +1,213 @@
+import logging
+import os
+import re
+from ast import literal_eval
+from collections.abc import Callable
+from functools import partial, update_wrapper
+
+import shapely
+from attrs import define
+from jsonschema import FormatChecker
+from jsonschema.validators import Draft202012Validator
+from referencing import Registry, Resource
+
+from harvester.records.exceptions import (
+    FieldValueInvalidWarning,
+    JSONSchemaValidationError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+###################
+# Data Validators
+###################
+class ValidateGeoshapeWKT:
+    """Decorator class for validating geoshape string values.
+
+    The validator should be applied to any field methods that retrieve geoshape
+    string values. If the field method retrieves a string value, the validator
+    will determine if shapely can parse the string as WKT.
+
+    Warning messages will be logged for the following scenarios:
+
+    1. Invalid WKT: Field method retrieves string value but shapely was unable to
+       parse string as WKT. Validator returns None.
+
+    2. Invalid None: Field method retrieves NoneType value. Validator returns None.
+
+    3. Invalid Type: Field method retrieves value that is neither of NoneType or a string
+       value. Validator returns None.
+
+    The validator will return the original value only if shapely was able to successfully
+    create a geometry object from the string value retrieved by the field method.
+    """
+
+    invalid_wkt_warning_message: str = (
+        "field: {field_name}, shapely was unable to parse string as WKT: '{value}'; "
+        "setting value to None"
+    )
+    invalid_none_warning_message: str = (
+        "field: {field_name}, value of type NoneType was provided; "
+        "returning original value of None"
+    )
+    invalid_type_warning_message: str = (
+        "field: {field_name}, value of type {type} was provided: {value}; "
+        "setting value to None"
+    )
+
+    def __init__(self, field_method: Callable):
+        update_wrapper(self, field_method)
+        self.field_method = field_method
+        self.field_name = field_method.__name__.removeprefix("_")
+
+    def __call__(self, obj: object) -> str | None:
+        """Validate string values retrieved by field method."""
+        value = self.field_method(obj)
+
+        if isinstance(value, str):
+            return self.validate(value)
+
+        if value is None:
+            logger.warning(
+                FieldValueInvalidWarning(
+                    self.invalid_none_warning_message.format(
+                        field_name=self.field_name, value=value
+                    )
+                )
+            )
+            return None
+
+        logger.warning(
+            FieldValueInvalidWarning(
+                self.invalid_type_warning_message.format(
+                    field_name=self.field_name, type=type(value), value=value
+                )
+            )
+        )
+        return None
```

Comment:
I might suggest a bit of simplifying here, namely just a shared, generic logging warning when the value validated is not a valid WKT string.  Something like:
```
field: {field_name}, unable to parse WKT from value: '{value}'; setting value to None"
```

This could apply to all scenarios:
- bad string (shapely fails)
- `None` value (potentially pretty common)
- non-string value passed.

For a couple of reasons:

- I think keeping the implementation detail of the `shapely` library out of the logging makes sense at the logging level; if we ever changed the validation logic or mechanism, we can continue to just say "it's not valid WKT"
- Not sure we need that granularity in the logging output, if they all return `None`, and this would reduce code complexity here

Not a required change, please pushback if you'd like to keep it!  Can also earmark for future optimization work that might occur around validations.

Reply:
Yes @jonavellecuerdo, looks good to me.  Apologies if I changed my thinking since our chat, or hadn't fully considered all the scenarios.  I'm recalling precisely what you mentioned now... that as a log reader I might wonder, "Why isn't this valid?  Looks good to me..." if I didn't know we ran it through `shapely` first.

I'm hopeful this decorator validation pattern is something we can apply to other fields in this project, and even other projects.  I think it's a fine line, and probably exceptions at every turn, but I think we should try and keep the **implementation details** (like, the lib `shapely` was used) out of the warning messages when something is not valid.  

We could optionally write another `DEBUG` line that gives more detail if we wanted?

But I think this looks good.

Reply:
Agreed

Reply:
@ghukill Your last comment makes a really good point! No apologies needed--lots of good learning along the way. 🤓 

Please see the latest commit!

Reply:
Thanks for your willingness to go back and forth a bit @jonavellecuerdo.  I think that last comment was the salient one; took awhile to get there.